postgres_replicator/transfer.go

```diff
@@ -0,0 +1,228 @@
+package transfer
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/bigquery"
+	"github.com/leporo/sqlf"
+	"google.golang.org/api/iterator"
+)
+
+// Transfer provides functions to transfer data from BigQuery to PostgreSQL.
+type Transfer struct {
+	bq     *BigQueryClient
+	pg     *PostgresClient
+	config *TableConfig
+	ready  chan bool
+}
+
+// NewTransfer returns a new Transfer.
+func NewTransfer(bq *BigQueryClient, pg *PostgresClient, config *TableConfig) *Transfer {
+	transfer := &Transfer{
+		bq:     bq,
+		pg:     pg,
+		config: config,
+		ready:  make(chan bool, 1),
+	}
+	transfer.ready <- true
+	return transfer
+}
+
+// Run creates a tableTransfer goroutine for each table to be transfered.
+func (t *Transfer) Run() {
+	select {
+	case <-t.ready:
+		log.Println("Beginning transfer(s)")
+	default:
+		log.Println("Transfer(s) already in progress, skipping")
+		return
+	}
+
+	activeTransfers := 0
+	done := make(chan bool)
+
+	for _, dataset := range t.config.Datasets {
+		for _, table := range dataset.Tables {
+			go t.transferTable(dataset.Name, table.Name, table.DateField, done)
+			activeTransfers++
+		}
+	}
+	for activeTransfers > 0 {
+		<-done
+		activeTransfers--
+	}
+
+	log.Println("All transfers complete")
+	t.ready <- true
+}
+
+// RunContinuously continuously runs Transfer.Run, with sleepTimeInSecs between
+// transfers.
+func (t *Transfer) RunContinuously(sleepAfterTransferInSecs int) {
+	for {
+		t.Run()
+		log.Printf("Sleeping for %d seconds", sleepAfterTransferInSecs)
+		time.Sleep(time.Duration(sleepAfterTransferInSecs) * time.Second)
+	}
+}
+
+func (t *Transfer) transferTable(bigQueryDataset, tableName, dateField string, done chan bool) {
+	logger := NewLogger(tableName)
+
+	// Get the BigQuery table schema
+	bqSchema, err := t.bq.GetTableSchema(bigQueryDataset, tableName)
+	if err != nil {
+		logger.Errorf("Could not get BigQuery table schema: %v", err)
+		done <- true
+		return
+	}
+
+	// Convert BigQuery schema to Postgres schema
+	pgSchema, err := t.convertSchema(bqSchema)
+	if err != nil {
+		logger.Errorf("Could not convert schema: %v", err)
+		done <- true
+		return
+	}
+
+	// Create PostgreSQL table if needed
+	err = t.prepareTable(tableName, pgSchema)
+	if err != nil {
+		logger.Errorf("Could not prepare Postgres table: %v", err)
+		done <- true
+		return
+	}
+
+	// Get rows to transfer
+	rows, err := t.getBigQueryRows(bigQueryDataset, tableName, dateField, bqSchema)
+	if err != nil {
+		logger.Errorf("Could not get data from BigQuery: %v", err)
+		done <- true
+		return
+	}
+
+	// Transfer rows to Postgres
+	err = t.transferToPostgres(tableName, pgSchema, rows, logger)
+	if err != nil {
+		logger.Errorf("Could not transfer one or more rows to Postgres: %v. ", err)
+		done <- true
+		return
+	}
+
+	done <- true
+}
+
+func (t *Transfer) convertSchema(bqSchema *BigQuerySchema) (*PostgresSchema, error) {
+	pgSchema := &PostgresSchema{make(map[string]string)}
```

Comment:
there should be a comment that explains the logic.
so you take the orignal schema as is and update the data types that are named differently in biqquery?
You should probably also link to the list of supported data types in bigquery and postgres and mention what are all the types that are supported (I really doubt that this would support all the data types, so you should add some explanation in places where you made assumptions).

Reply:
Structs, arrays, etc are converted to JSON, which can then be queried / typecast from Postgres queries. I added a section to the README that explains this.